src/server/scripts/Kalimdor/CavernsOfTime/BattleForMountHyjal/instance_hyjal.cpp

```diff
@@ -132,128 +133,127 @@ class instance_hyjal : public InstanceMapScript
+                }
+                break;
+            case NPC_TOWERING_INFERNAL:
+            case NPC_LESSER_DOOMGUARD:
+            case NPC_DIRE_WOLF:
+            case NPC_GUARDIAN_ELEMENTAL:
+                if (creature->IsSummon())
+                {
+                    _summonedNPCs.insert(creature->GetGUID());
+                }
+                break;
+            }
+            InstanceScript::OnCreatureCreate(creature);
+        }
+
+        void OnUnitDeath(Unit* unit) override
+        {
+            InstanceScript::OnUnitDeath(unit);
+
+            if (Creature* creature = unit->ToCreature())
+            {
+                switch (creature->GetEntry())
+                {
                 case NPC_NECRO:
                 case NPC_ABOMI:
                 case NPC_GHOUL:
                 case NPC_BANSH:
                 case NPC_CRYPT:
-                case NPC_STALK:
                 case NPC_GARGO:
                 case NPC_FROST:
                 case NPC_INFER:
-                    if (_bossWave != TO_BE_DECIDED)
-                        creature->AI()->DoAction(_bossWave);
-                    else if (_retreat)
-                        creature->AI()->DoAction(_retreat);
-
-                    if (creature->IsSummon() && _bossWave != TO_BE_DECIDED)
+                case NPC_STALK:
+                    if (creature->IsSummon())
                     {
-                        if (_currentWave == 0)
-                            creature->SetDisableReputationGain(true);
-                        DoUpdateWorldState(WORLD_STATE_ENEMYCOUNT, ++trash);    // Update the instance wave count on new trash spawn
-                        _encounterNPCs.insert(creature->GetGUID());             // Used for despawning on wipe
+                        if (_bossWave != TO_BE_DECIDED)
+                        {
+                            DoUpdateWorldState(WORLD_STATE_ENEMYCOUNT, --_trash);    // Update the instance wave count on new trash death
+                            _encounterNPCs.erase(creature->GetGUID());    // Used for despawning on wipe
+
+                            if (_trash == 0) // It can reach negatives if trash spawned after a retreat are killed, it shouldn't affect anything. Also happens on retail
+                                SetData(DATA_SPAWN_WAVES, 1);
+                        }
                     }
                     break;
                 case NPC_TOWERING_INFERNAL:
                 case NPC_LESSER_DOOMGUARD:
                 case NPC_DIRE_WOLF:
                 case NPC_GUARDIAN_ELEMENTAL:
-                    if (creature->IsSummon())
-                    {
-                        _summonedNPCs.insert(creature->GetGUID());
-                    }
+                    _summonedNPCs.erase(creature->GetGUID());
+                    break;
+                case NPC_WINTERCHILL:
+                    _initialWaves = false;
+                    ResetWaves();
+                    break;
```

Comment:
I see you extracted and created `void ResetWaves()` to avoid duplication. This feels to me like a good place to use `[[fallthrough]]` instead
```suggestion
                   [[fallthrough]];
```